### PR TITLE
Return an error rather than panicking on a bad request.

### PR DESCRIPTION
**Pull request by connor4312**
It's considered much more idiomatic to return errors rather than panicking if something goes wrong. It might actually be _dangerous_ to panic here... the expectation is that if there's a possiblity for an error, an error will be returned. If consumers don't see that an error is returned, they may simply assume `nil` is returned on an invalid request, and applications will burn...

Unfortunately this is not a backwards-compatible change.

### CODE REVIEW

**Comment by eapache** [comment]
I deliberately chose to panic here (being well aware of the idiomatic style) for a handful of reasons I don't want to try typing up on a phone while on vacation. I will provide more explanation next week.

**Comment by eapache** [comment]
Longer answer is:

This package was originally extracted from another (https://github.com/eapache/channels/). In that package, where this was a strictly internal component, any of the error cases indicated a bug in the other code in that package, so a panic was the correct thing to do since none of the state could be trusted at that point.

Once extracted, the API was designed for that use case (thus e.g. the fact I never bothered to implement `Pop()`) and the panics remained. All the other cases I've ended up using it (and seeing it used) have followed a similar pattern, where a queue-level error case indicated a critical state bug in the calling code, not something that should be returned to the user as a normal error.

I grant that there are other use cases where it may make more sense to return errors, but those weren't my original target, and I'm not sure it makes sense to try and support both in a single package. Perhaps there is a phrase I could add in the godoc to make this intention clearer? The fact of the panics themselves is already documented appropriately.

---

TL;DR, I'm not going to accept this PR, but thanks for giving me an excuse to explain my reasoning :)

I may or may not accept a PR adding a side-by-side error-returning implementation, I haven't given that a ton of thought. On one hand, there are definitely use cases where it makes sense. On the other hand, the duplication isn't great.

**Comment by connor4312** [comment]
Ah, thanks, that makes sense :)
